Build and release processes improvements

FabienLelaquais opened the pull request:
Addresses #2496, #2512 and hopefully more.

- Package dependencies fixed:
    M.m.t depends on >=M.m,<M.(m+1)
- 'taipy' is now a first-class citizen for single-package build
- Enforce Python <3.13 in all pyprojet.toml
- Workflows UI simplified.
  + Version syntax checked.
- Upgrade codeql-action version.
- Release-dedicated scripts:
  - Added comments and command line argument checks.
  - Added tests on release-dedicated scripts
  - Added Package and Version classes for more code reuse.
  - Fixed missing versions fetched from Pypi.

FabienLelaquais committed:
- Package dependencies fixed:
    M.m.t depends on >=M.m.t,<M.(m+1)
- Starting work on making 'taipy' a first-class citizen for single-package build
- Enforce Python <3.13 in all pyprojet.toml
- Workflows UI simplified.
  + Version syntax checked.
- Upgrade codeql-action version.
- Release-dedicated scripts:
  - Added comments and command line argument checks.
  - Added tests on release-dedicated scripts
  - Added Package and Version classes for more code reuse.
  - Fixed missing versions fetched from Pypi.

FabienLelaquais committed:
Fix test

FabienLelaquais committed:
Fix test (2)

jrobinAV commented on tools/release/check_package_version.py:
```diff
@@ -0,0 +1,41 @@
+
+def usage() -> None:
+    print(f"Usage: {sys.argv[0]} <package> <version>")  # noqa: T201
+    print("   Checks that the <package>'s version.json file <version>.")  # noqa: T201
```
```suggestion
    print("   Checks the <package>'s version.json file <version>.")  # noqa: T201
```

or 

```suggestion
    print("   Checks that the version in the <package>'s version.json file matches the <version> input.")  # noqa: T201
```

jrobinAV commented on tools/release/check_package_version.py:
rename to check_package_version_equals_input_version.py

FabienLelaquais replied:
Won't do. I really hate this long names.
+ That's the original file name... the script checks the version of the package.

FabienLelaquais committed:
- Favor pyproject vs setup (simplified MANIFEST.in).
- Optimized taipy package build.
- Adapt sub-packages compatibility rules.
- Added version check on frontend version.
- GH workflows can run from any remote

FabienLelaquais committed:
Linter

FabienLelaquais committed:
Merge branch 'release/4.0' into bug/#2496-build-process-limitations

FabienLelaquais committed:
Linter (2)

FabienLelaquais committed:
Merge branch 'bug/#2496-build-process-limitations' of github.com:Avaiga/taipy into bug/#2496-build-process-limitations

FabienLelaquais committed:
- Generated dependencies on sub-pacakges now are >=M.m,<M.(n+1) instead of >=M.m.t
- Allow to force a dependency version in the dependent's setup.requirements.txt file.

jrobinAV commented on tests/release/test_version.py:
❤️

jrobinAV commented on tests/release/test_version.py:
What about moving it to `tests/tools/release/test_version.py`

jrobinAV commented on tools/release/update_setup_requirements.py:
```diff
@@ -8,45 +8,87 @@
+                usage()
+                raise ValueError("Couldn't figure out GitHub branch path.")
+        else:
+            gh_path = sys.argv[len(PACKAGES) + 3]
```
I don't think iterating over a list of sub-packages instead of hard coding the list brings value. On the contrary, the code is now very difficult to read, understand, and then maintain. If it does the same, I would personally revert it.

jrobinAV commented on tools/release/update_setup_requirements.py:
```diff
@@ -8,45 +8,87 @@
+    if use_pypi:
+        # Target dependency version should the latest compatible with 'version'
+        major, minor = version.split(".")[:2]
+        return f"{line} >={major}.{minor},<{major}.{int(minor) + 1}\n"
```
👍

jrobinAV commented on tools/release/update_setup_requirements.py:
```diff
@@ -8,45 +8,87 @@
+        usage()
+        raise ValueError("Missing arguments.")
+    package = sys.argv[1]
+    versions = {f"taipy-{p}": sys.argv[i] for i, p in enumerate(PACKAGES, 2)}
```
Why don't we instantiate Version objects instead of using string?

jrobinAV commented on tools/release/update_setup_requirements.py:
```diff
@@ -8,45 +8,87 @@
-    _path = os.path.join(BASE_PATH, package, "setup.requirements.txt")
+
+def update_setup_requirements(
+    package: str, versions: dict[str, str], publish_on_py_pi: bool, gh_path: t.Optional[str]
```
Why not using the Version class?
```suggestion
    package: str, versions: dict[str, Version], publish_on_py_pi: bool, gh_path: t.Optional[str]
```

jrobinAV commented on tools/release/update_setup_requirements.py:
```diff
@@ -8,45 +8,87 @@
+        print(f"Generated setup.requirements.txt for package '{package}'")  # noqa: T201
+        for line in lines:
+            print(line.strip())  # noqa: T201
+        print("-" * 32)  # noqa: T201
```
I guess you wanted to prepare the future to make this piece of code de-activable. I would not. It does not cost much to have it all the time and it can always be useful.

```suggestion
    # More logs
    print(f"Generated setup.requirements.txt for package '{package}'")  # noqa: T201
    for line in lines:
        print(line.strip())  # noqa: T201
    print("-" * 32)  # noqa: T201
```

jrobinAV commented on .github/workflows/build-and-release-single-package.yml:
```diff
@@ -130,41 +132,41 @@ jobs:
+      - name: Build Taipy front-end
+        if: ${{ github.event.inputs.target_package == 'taipy' }}
         run: |
-          cp -r tools/packages/taipy-${{ github.event.inputs.target_package }}/. ${{ steps.set-variables.outputs.package_dir }}
```
So this copy thing is not done anymore. What was the purpose of that?

jrobinAV commented on .github/workflows/build-and-release-single-package.yml:
```diff
@@ -130,41 +132,41 @@ jobs:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
       - name: Ensure Taipy release is marked as latest
         run: |
```
We only want to set the release as latest if it is the taipy package.

```suggestion
        if: github.event.inputs.target_package == 'taipy'
        run: |
```

jrobinAV commented on .github/workflows/publish.yml:
```diff
@@ -8,6 +8,16 @@ on:
         required: true
 
 jobs:
+  check-version:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Validate version input
+        run: |
+          if [[ ! "${{ github.event.inputs.version }}" =~ ^[0-9]+\.[0-9]+\.[0-9]+(\.[a-zA-Z]+[0-9]*)?$ ]]; then
+            echo "❌ Invalid version format. Expected format: M.m.t[.ext]"
+            exit 1
+          fi
+
```
I believe this is already done later in a Python script.

jrobinAV commented on taipy/common/pyproject.toml:
```diff
@@ -22,21 +22,11 @@ classifiers = [
-    "taipy.common._cli",
-    "taipy.common._cli.*"
-]}
+find = {include = ["taipy", "taipy.common", "taipy.common.*"]}
```
I think I remember we hardcoded the subfolders on purpose. But I don't recall why.

If someone remembers, please jump on.

jrobinAV commented on taipy/common/version.json:
```diff
@@ -1 +1 @@
-{"major": 4, "minor": 0, "patch": 2}
+{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}
```
```suggestion
{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}

```

jrobinAV commented on taipy/core/version.json:
```diff
@@ -1 +1 @@
-{"major": 4, "minor": 0, "patch": 2}
+{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}
```
```suggestion
{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}

```

jrobinAV commented on taipy/gui/version.json:
```diff
@@ -1 +1 @@
-{"major": 4, "minor": 0, "patch": 3, "ext": "dev1"}
+{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}
```
```suggestion
{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}

```

jrobinAV commented on taipy/rest/version.json:
```diff
@@ -1 +1 @@
-{"major": 4, "minor": 0, "patch": 2}
+{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}
```
```suggestion
{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}

```

jrobinAV commented on taipy/templates/version.json:
```diff
@@ -1 +1 @@
-{"major": 4, "minor": 0, "patch": 2}
+{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}
```
```suggestion
{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}

```

jrobinAV commented on taipy/version.json:
```diff
@@ -1 +1 @@
-{"major": 4, "minor": 0, "patch": 3, "ext": "dev1"}
+{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}
```
```suggestion
{"major": 4, "minor": 0, "patch": 3, "ext": "dev0"}

```

jrobinAV commented on frontend/taipy-gui/src/components/Taipy/TableSort.tsx:
```diff
@@ -136,7 +136,7 @@ const SortRow = (props: SortRowProps) => {
             <Grid size={6}>
                 <FormControl margin="dense">
-                    <InputLabel>Column</InputLabel>
+                    <InputLabel>{fieldHeader}</InputLabel>
```
hum?

jrobinAV commented on tools/release/build_package_structure.py:
I did not review this. but I trust you!

jrobinAV commented on tools/release/common.py:
```diff
@@ -0,0 +1,214 @@
+# These are the base name of the sub packages taipy-*
+# They also are the names of the directory where their code belongs, under the 'taipy' directory
+# in the root of the Taipy repository.
+PACKAGES = ["common", "core", "gui", "rest", "templates"]
```
Minor improvement: We could directly expose the Package instances.

jrobinAV reviewed (dismissed):
Question: 
Don't we want to merge it to the develop branch first, so we can test it with a dev version?

FabienLelaquais commented:
> Question: Don't we want to merge it to the develop branch first, so we can test it with a dev version?

Well... that really depends on how bad the merge could turn out to be...

FabienLelaquais committed:
- Fix 'latest' release to latest taipy's.
- JR comments

jrobinAV commented on .github/workflows/build-and-release-single-package.yml:
```diff
@@ -130,41 +132,42 @@ jobs:
 
-      - name: Ensure Taipy release is marked as latest
+      - name: Ensure 'taipy' release is marked as latest
+        if: ${{ github.event.inputs.target_package != 'taipy' || github.event.inputs.release_type == 'production' }}
```
```suggestion
        if: ${{ github.event.inputs.target_package != 'taipy' || github.event.inputs.release_type != 'production' }}
```

jrobinAV commented on .github/workflows/build-and-release-single-package.yml:
```diff
@@ -130,41 +132,42 @@ jobs:
+        if: ${{ github.event.inputs.target_package != 'taipy' || github.event.inputs.release_type == 'production' }}
         run: |
-           gh release edit ${{needs.fetch-versions.outputs.taipy_VERSION}} --latest
+          gh release edit ${{ needs.fetch-versions.outputs.taipy_VERSION }} --latest
```
I don't see where or how `taipy_VERSION` is set. So I don't know if that works.

jrobinAV commented on frontend/taipy-gui/src/components/Taipy/TableSort.tsx:
```diff
@@ -248,7 +248,7 @@ const TableSort = (props: TableSortProps) => {
                 open={showSort}
                 onClose={onShowSortClick}
-                className={getSuffixedClassNames(className, "-filter")}
+                className={getSuffixedClassNames(className, "-sort")}
```
😮 

It does not seem related to the current schmilblick.

jrobinAV commented on tools/release/setup_version.py:
```diff
@@ -13,86 +13,86 @@
-    for _path in paths:
-        _version = extract_version(_path)
-        _name = None if _path == "taipy" else _path.split(os.sep)[-1]
+    for package_name in packages:
+        package = Package(package_name)
+        version = extract_version(package)
```
❤️

FabienLelaquais committed:
Fix for issue#2512

FabienLelaquais committed:
Linter

FabienLelaquais committed:
Keep latest when release type is dev

FabienLelaquais committed:
Bump taipy version
